# TensorFlow and Keras 

In [16]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt #imshow = 'image show'
from sklearn.linear_model import Perceptron
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import tensorflow as tf


%matplotlib inline

In [4]:
print(tf.__version__)

2.16.1


In [ ]:
rank_0 = 

In [10]:
rank_1_t = tf.constant([1,2,3])
rank_1_t

<tf.Tensor: shape=(3,), dtype=int32, numpy=array([1, 2, 3])>

In [6]:
rank_2_t = tf.constant(
    [
        [5,8,9],
        [11,2,6]
    ]
    , dtype='float64'
)
rank_2_t

<tf.Tensor: shape=(2, 3), dtype=float64, numpy=
array([[ 5.,  8.,  9.],
       [11.,  2.,  6.]])>

In [8]:
#main braket 
    #bracket for each sub component
rank_3_t = tf.constant([
    [                       #1st component
        [0,4,5,6,8],
        [1,2,3,4,5]
    ],
    [                       #1st component
        [6,8,1,2,0],
        [2,3,4,4,4]
    ],
    [                       #1st component
        [0,0,0,1,2],
        [9,9,9,0,9]
    ]
])
rank_3_t

<tf.Tensor: shape=(3, 2, 5), dtype=int32, numpy=
array([[[0, 4, 5, 6, 8],
        [1, 2, 3, 4, 5]],

       [[6, 8, 1, 2, 0],
        [2, 3, 4, 4, 4]],

       [[0, 0, 0, 1, 2],
        [9, 9, 9, 0, 9]]])>

### convert dataframe into tensor 

In [11]:
data = {
    'age': [4,5,6,7],
    'score':[30,35,40,24]
}
df = pd.DataFrame(data)
df

age  score
0    4     30
1    5     35
2    6     40
3    7     24

In [13]:
tensor = tf.convert_to_tensor(df.values,dtype='float32')
tensor

<tf.Tensor: shape=(4, 2), dtype=float32, numpy=
array([[ 4., 30.],
       [ 5., 35.],
       [ 6., 40.],
       [ 7., 24.]], dtype=float32)>

In [ ]:
#reading csv directly as tensor

#option1 
dataset = tf.data.TextLineDataset('myfile.csv')

#option2 
dataset = tf.convert_to_tensor(pd.read_csv('my_file.csv'))

# Tensor Arithmetic

In [25]:
a = tf.constant([[1,2],[3,4]])
b = tf.constant([[5,6],[7,8]])

print(tf.add(a,b))
print(tf.multiply(a,b))

tf.Tensor(
[[ 6  8]
 [10 12]], shape=(2, 2), dtype=int32)
tf.Tensor(
[[ 5 12]
 [21 32]], shape=(2, 2), dtype=int32)


In [17]:
a[0]

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([1, 2])>

### additional functions that are used in deep learning 


In [18]:
print(tf.reduce_max(b)) #grabs the maximum value of the tensor 

tf.Tensor(8, shape=(), dtype=int32)


In [19]:
print(tf.math.argmax(b))

tf.Tensor([1 1], shape=(2,), dtype=int64)


In [26]:
x = tf.constant([
    [3.0,5.2],
    [6.3,8.6]
])

print(tf.nn.softmax(x)) #needs decimal numbers 

tf.Tensor(
[[0.0997505  0.90024954]
 [0.09112294 0.9088771 ]], shape=(2, 2), dtype=float32)


In [28]:
# Ragged Tensor: 
ragged_list = [
    [4,5,6,7],
    [4,6],
    [7,8,9,12,122,43,4],
    [4]
]

ragged_tensor = tf.ragged.constant(ragged_list) #building tensors for text data since it has irregular lengths 
print(ragged_tensor)

<tf.RaggedTensor [[4, 5, 6, 7], [4, 6], [7, 8, 9, 12, 122, 43, 4], [4]]>


# Keras
- app built on top of tensorflow 
- developed by Francois Chollet, a google AI engineer 
- it's fast, easy to implement, and modular in nature (it's like the seaborn to matplotlib)
- written in python 



In [1]:
from tensorflow.keras import models, layers

#import the dataset 
from tensorflow.keras.datasets import mnist 

#for viz 
import matplotlib.pyplot as plt

In [2]:
#extract the dataset and split it 
(x_train,y_train),(x_test,y_test) = mnist.load_data()

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 24s 2us/step


In [3]:
x_train #same as peceptron BUT not flattened here 

array([[[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       ...,

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]],

       [[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 

In [6]:
#check the value range
print(f'range: {x_train.min()},{x_train.max()}')

range: 0,255


In [7]:
x_train.shape

(60000, 28, 28)

In [8]:
#data normalization. since we know x ranges from 0-255, we can divide by 255 so that all numbers are between 0 and 1 
x_train,x_test = x_train/255.0, x_test/255.0 #add the '.0' to preserve the decimal 


In [9]:
#check the value range
print(f'range: {x_train.min()},{x_train.max()}')

range: 0.0,1.0


## Building the neural network model 

In [10]:
nn_model = models.Sequential(  #model building happens sequentially
    [layers.Flatten(input_shape=(28,28)),  #first layer is the input layer with flatten function 
    layers.Dense(128,activation='relu'), #model with 1 hidden layer with 128 neurons (choosen any number; takes a lot of trial and error to find the right number for your dataset) with ReLU activation function
    layers.Dropout(0.15), #apply regularization: also needs hyperparameter tuning. need to do regularization per layer. this is to address overfitting
    layers.Dense(10)  #output layer: need 10 neurons since we want to predict 10 outcomes (digits 0-9)
    ]
)   


c:\Users\chand\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


In [12]:
nn_model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten (Flatten)               │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 128)            │       100,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 128)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 101,770 (397.54 KB)

 Trainable params: 101,770 (397.54 KB)

 Non-trainable params: 0 (0.00 B)

- 100,480 parameters = 784 inputs (x) and 128 weights and 128 biases (784*128+128)
- 1290 parameters = 128 neuron outputs (weights)*10 output neurons + 10 biases 

In [14]:
x_train[:1]

array([[[0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        ],
        [0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        ],
        [0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.        , 0.        , 0.        , 0.        , 0.        ,
         0.

In [13]:
prediction = nn_model(x_train[:1])
prediction

<tf.Tensor: shape=(1, 10), dtype=float32, numpy=
array([[-0.10240273,  0.05184153,  0.26182458, -0.47021058, -0.09501468,
         0.08849951,  0.5627961 ,  0.05309828, -0.11336643,  0.39039654]],
      dtype=float32)>

In [17]:
#need an activation function to read the prediction output 
#apply softmax, which is used when there is multiple classes i.e. predicting 10 outcomes, 0-9 
tf.nn.softmax(prediction)
#these are the predictions or probability of the image being each number. We hope to see one of these numbers to be higher than 80% and the rest to be low numbers. Note the prediction probability always adds up to 1


<tf.Tensor: shape=(1, 10), dtype=float32, numpy=
array([[0.0815974 , 0.09520588, 0.11745143, 0.05648582, 0.08220248,
        0.09876069, 0.15869693, 0.0953256 , 0.08070768, 0.13356613]],
      dtype=float32)>

In [18]:
tf.nn.softmax(prediction).numpy()[0].sum()

1.0

In [ ]:
#we can see that the above probabilities is indicisive. we'll try back propagation next. 